src/main/scala/dx/core/languages/cwl/CwlUtils.scala

```diff
@@ -0,0 +1,154 @@
+package dx.core.languages.cwl
+
+import dx.translator.CallableAttributes.{DescriptionAttribute, TitleAttribute}
+import dx.cwl.{ArrayValue, BooleanValue, CwlArray, CwlBoolean, CwlDirectory, CwlDouble, CwlEnum, CwlFile, CwlFloat, CwlInt, CwlLong, CwlNull, CwlOptional, CwlRecord, CwlString, CwlType, CwlValue, DirectoryValue, DoubleValue, FileValue, FloatValue, IntValue, LongValue, NullValue, ObjectValue, StringValue}
+import dx.core.ir.{CallableAttribute, ParameterAttribute, Type, Value}
+import dx.core.ir.Type.{TArray, TBoolean, TDirectory, TEnum, TFile, TFloat, THash, TInt, TOptional, TSchema, TString}
+import dx.core.ir.Value.{VArray, VBoolean, VDirectory, VFile, VFloat, VHash, VInt, VNull, VString}
+import dx.translator.ParameterAttributes.{HelpAttribute, LabelAttribute}
+import wdlTools.syntax.SourceLocation
+
+object CwlUtils {
+  val locPlaceholder: SourceLocation = SourceLocation.empty
+
+  def toIRType(cwlTypes: Vector[CwlType]): Type = {
+    cwlTypes match {
+      case _ if cwlTypes.size > 1 && cwlTypes.contains(CwlNull) => TOptional(toIRType(cwlTypes.filter(_ != CwlNull)))
+      case _ if cwlTypes.length == 1 => cwlTypes.head match {
+        case CwlBoolean => TBoolean
+        case CwlInt => TInt
+        case CwlLong => TInt
+        case CwlDouble => TFloat
+        case CwlFloat => TFloat
+        case CwlString => TString
+        case CwlFile => TFile
+        case CwlDirectory => TDirectory
+        case a: CwlArray => TSchema(a.name.get, Map[String, Type]("array" -> TArray(toIRType(a.itemTypes))))
+        case r: CwlRecord => TSchema(r.name.get, r.fields.map({ case (key, value) => key -> toIRType(value.types) }))
+        case e: CwlEnum => TSchema(e.name.get, Map[String, Type]("enum" -> TEnum(e.symbols)))
+        case _ => throw new Exception(s"Cannot convert CWL type ${cwlTypes.head} to IR")
+      }
+      case _ if cwlTypes.isEmpty => throw new IllegalArgumentException("Vector of CwlTypes must be non-empty!")
+      case other => throw new NotImplementedError(s"Multiple types are not supported yet! (${other} was given)")
+    }
+  }
+
+  def fromIRType(irType: Type): CwlType = {
+    def inner(innerType: Type): CwlType = {
```

Comment:
Use of an inner function isn't necessary here. I only use this pattern when the outer function has some additional arguments that I don't want to have to pass to the recursive call every time.